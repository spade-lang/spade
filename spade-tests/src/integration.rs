#[cfg(test)]
mod namespace_tests {
    use crate::{build_items, snapshot_error};

    #[test]
    fn namespacing_works() {
        let code = r#"
            mod X {
                entity x() -> int<2> {
                    1
                }
            }

            entity top() -> int<2> {
                inst X::x()
            }
        "#;

        build_items(code);
    }

    snapshot_error!(
        namespacing_adds_to_the_correct_namespace,
        r#"
            mod X {
                entity x() -> int<2> {
                    1
                }
            }

            entity top() -> int<2> {
                x()
            }
        "#
    );

    #[test]
    fn use_statements_work() {
        let code = r#"
            mod X {
                entity x() -> int<2> {
                    1
                }
            }

            use X::x;

            entity top() -> int<2> {
                inst x()
            }
            "#;

        build_items(code);
    }

    #[test]
    fn renaming_use_statements_work() {
        let code = r#"
            mod X {
                entity x() -> int<2> {
                    1
                }
            }

            use X::x as a;

            entity top() -> int<2> {
                inst a()
            }
            "#;

        build_items(code);
    }

    /// NOTE This test fails currently
    #[test]
    fn recursive_use_statements_work() {
        let code = r#"
            mod X {
                mod Y {
                    entity x() -> int<2> {
                        1
                    }
                }
                use Y::x;
            }

            use X::x as a;

            entity top() -> int<2> {
                inst a()
            }
        "#;

        build_items(code);
    }

    #[test]
    fn using_names_in_namespaces_works() {
        let code = r#"
            mod X {
                enum A {X{a: bool}}

                entity x() -> A {
                    A::X(true)
                }
            }
            "#;

        build_items(code);
    }

    #[test]
    fn using_names_of_types_in_namespaces_works() {
        let code = r#"
            mod X {
                struct A {}
                struct B{a: A}
            }
            "#;

        build_items(code);
    }

    // NOTE: this is an awful error message at the moment, but it is strange code
    // and fixing it would take quite a bit of effort, so we'll leave it be and
    // create an issue for it
    snapshot_error! {
        pipeline_shadowing_does_not_fail_silently,
        "
        pipeline(2) main(clk: clk, x: int<8>) -> int<8> {
                let x: int<8> = 0;
            reg;
                let x: int<8> = 1;
            reg;
                stage(-2).x
        }
        "
    }

    snapshot_error! {
        backward_types_can_not_be_put_in_registers,
        "
        entity x(clk: clk, a: ~bool) -> bool {
            reg(clk) _ = a;
            true
        }
        "
    }

    snapshot_error! {
        transitive_backwar_type_can_not_be_put_in_registers,
        "
        struct X {
            a: ~bool,
            b: bool
        }
        entity x(clk: clk, a: X) -> bool {
            reg(clk) _ = a;
            true
        }
        "
    }
}
