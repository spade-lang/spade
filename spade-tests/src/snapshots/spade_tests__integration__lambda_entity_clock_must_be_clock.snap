---
source: spade-tests/src/integration.rs
---
entity test(clk: clock) {
    let _ = entity (clk) {
        let clk: uint<8> = clk;
    }.inst call(clk, ());
}


error: Expected type uint<8>, got clock
  ┌─ testinput:3:28
  │
2 │     let _ = entity (clk) {
  │                     --- Type clock inferred here
3 │         let clk: uint<8> = clk;
  │                  -------   ^^^ Expected uint<8>
  │                  │          
  │                  Type uint<8> inferred here
  │
  = note: Expected: uint<8>
               Got: clock
  = The error is in a generic unit instantiated at
    ╰ testinput:2 impl_12::call<()>
